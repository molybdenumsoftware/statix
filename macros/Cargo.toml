[package]
edition.workspace = true
license.workspace = true
name = "macros"
version = "0.0.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
hex.workspace = true
proc-macro2.workspace = true
quote.workspace = true
sha2.workspace = true
syn.workspace = true

[lib]
proc-macro = true

[lints]
workspace = true
