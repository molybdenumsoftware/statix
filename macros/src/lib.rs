mod explain;
mod metadata;

use explain::generate_explain_impl;
use metadata::{generate_meta_impl, RawLintMeta};
use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;
use quote::quote;
use syn::{parse_macro_input, Ident, ItemStruct};

fn generate_self_impl(struct_name: &Ident) -> TokenStream2 {
    quote! {
        impl #struct_name {
            pub fn new() -> Self {
                Self
            }
        }
    }
}

#[proc_macro_attribute]
pub fn lint(attr: TokenStream, item: TokenStream) -> TokenStream {
    let struct_item = parse_macro_input!(item as ItemStruct);
    let meta = parse_macro_input!(attr as RawLintMeta);

    let struct_name = &struct_item.ident;
    let self_impl = generate_self_impl(struct_name);
    let meta_impl = generate_meta_impl(struct_name, &meta);
    let explain_impl = generate_explain_impl(&struct_item);

    (quote! {
        #struct_item

        ::lazy_static::lazy_static! {
            pub static ref LINT: Box<dyn crate::Lint> = Box::new(#struct_name::new());
        }

        #self_impl
        #meta_impl
        #explain_impl

        impl crate::Lint for #struct_name {}
    })
    .into()
}
