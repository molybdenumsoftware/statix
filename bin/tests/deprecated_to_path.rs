mod _utils;

use macros::generate_tests;

generate_tests! {
    rule: manual_inherit,
    expressions: [
        "builtins.toPath x",
        "toPath x",
        r#"toPath "/abc/def""#,
        r#"builtins.toPath "/some/path""#,
    ],
}
