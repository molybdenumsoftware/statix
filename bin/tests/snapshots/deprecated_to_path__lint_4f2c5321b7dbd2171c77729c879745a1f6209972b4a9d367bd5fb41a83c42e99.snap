---
source: bin/tests/deprecated_to_path.rs
expression: "\"toPath x\""
---
[W17] Warning: Found usage of deprecated builtin toPath
   ╭─[<temp_file_path>:1:1]
   │
 1 │ toPath x
   · ────┬───  
   ·     ╰───── toPath is deprecated, see :doc builtins.toPath within the REPL for more
───╯
